File: src/p327/Solution.java
package p327;

import java.util.Arrays;

class SegmentTree {
	long nums[];
	int n;
	int tree[];
	int count[];
	int left[];
	int right[];

	public SegmentTree(long nums[]) {
		this.nums = nums;
		this.n = nums.length;
		buildTree();
	}

	public void buildTree() {
		int treeSize = (n << 2) + 4;
		tree = new int[treeSize];
		count = new int[treeSize];
		left = new int[treeSize];
		right = new int[treeSize];
		buildTree(0, 0, n - 1);
	}

	private void buildTree(int node, int left, int right) {
		this.left[node] = left;
		this.right[node] = right;
		if (left != right) {
			int mid = ((left + right) >> 1);
			buildTree(node + node + 1, left, mid);
			buildTree(node + node + 2, mid + 1, right);
		}
	}

	public void addNumber(long num) {
		addNumber(0, num);
	}

	private void addNumber(int node, long num) {
		count[node]++;
		if (left[node] == right[node])
			return;
		if (num <= nums[(left[node] + right[node]) >> 1])
			addNumber(node + node + 1, num);
		else
			addNumber(node + node + 2, num);
	}

	public int query(long lower, long upper) {
		return query(0, lower, upper);
	}

	private int query(int node, long lower, long upper) {
		if (nums[right[node]] <= upper && nums[left[node]] >= lower)
			return count[node];
		if (nums[right[node]] < lower || nums[left[node]] > upper)
			return 0;
		return query(node + node + 1, lower, upper) + query(node + node + 2, lower, upper);
	}
}

public class Solution {
	public int countRangeSum(int[] nums, int lower, int upper) {
		if(nums.length == 0)
			return 0;
		long sums[] = new long[nums.length + 1];
		sums[0] = nums[0];
		sums[nums.length] = 0;
		for (int i = 1; i < nums.length; i++)
			sums[i] = sums[i - 1] + nums[i];
		long[] sortedSums = Arrays.copyOf(sums, sums.length);
		Arrays.sort(sortedSums);
		SegmentTree tree = new SegmentTree(sortedSums);
		int ans = 0;
		tree.addNumber(0);
		for (int i = 0; i < nums.length; i++) {
			ans += tree.query(sums[i] - upper, sums[i] - lower);
//			System.out.println(ans);
			tree.addNumber(sums[i]);
		}
		return ans;
	}
/*
 * [2147483647,-2147483648,-1,0]
-1
0
[-4,0,-3,-1,1,2,1,-4]
0
6
 */
	public static void main(String[] args) {
		int array[] = { -4,0,-3,-1,1,2,1,-4 };
		System.out.println(new Solution().countRangeSum(array, 0, 6));
	}
}

File: src/p327/solution.py
class Solution(object):
    def merge(self, nums, left, mid, right):
        result = 0
        i, j = left, mid + 1
        lower_result, upper_result = 0, 0
        while i <= mid and j <= right:
            if nums[i] < nums[j] - self.upper:
                lower_result += j - mid - 1
                i += 1
            else:
                j += 1
        while i <= mid:
            lower_result += right - mid
            i += 1
        i, j = left, mid + 1
        while i <= mid and j <= right:
            if nums[i] <= nums[j] - self.lower:
                upper_result += j - mid - 1
                i += 1
            else:
                j += 1
        while i <= mid:
            upper_result += right - mid
            i += 1
        result = lower_result - upper_result
        i, j = left, mid + 1
        l = []
        while i <= mid and j <= right:
            if nums[i] <= nums[j]:
                l.append(nums[i])
                i += 1
            else:
                l.append(nums[j])
                j += 1
        while i <= mid:
            l.append(nums[i])
            i += 1
        while j <= right:
            l.append(nums[j])
            j += 1
        assert len(l) == right - left + 1
        for i in range(len(l)):
            nums[i + left] = l[i]
        return result

    def mergesort(self, nums, left, right):
        if left >= right:
            return 0
        mid = (left + right) >> 1
        result = 0
        result += self.mergesort(nums, left, mid)
        result += self.mergesort(nums, mid + 1, right)
        result += self.merge(nums, left, mid, right)
        return result

    def countRangeSum(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: int
        """
        sums = list(nums)
        for i in range(1, len(nums)):
            sums[i] += sums[i - 1]
        self.lower = lower
        self.upper = upper
        return self.mergesort([0] + sums, 0, len(sums))
